### ISSUE TITLE
User-selectable fee

### ISSUE THREAD

**Issue by keo** [voisine/breadwallet#384]
Hi,

we're running an ATM and an online exchange and again we have a great influx of support requests because people can't cash out.

Here's an example support request: "So I got my BTC now in the breadwallet, is there any way to move it quickly out of that wallet to a wallet where I can set fees?"

We need 1 confirmation for the ATM to give out cash above 100 EUR. And there is a 6 hour time window to do it.

When the network is congested, breadwallet users' transaction usually confirm in 8-16 hours, which means we have to do everything by hand for all these stuck transactions (giving them cash, covering on an exchange, doing bookkeeping, clearing the database, etc.)

**It would be great if there would be a way (somewhere deep and hidden in some kind of advanced settings)  to set a fee priority level.**

We're still recommending breadwallet to all our users because we love it but we're on the brink of looking for other alternatives as this has a huge load on our support staff.

Thank you for reconsidering.

**Comment by dabura667** [comment]
Would be interesting if there was an attribute added to the bitcoin URI which would be understood as meaning "this payment relies on getting confirmations quickly"

ie.

bitcoin:1376dsfksuhefkdskdsjdsksdfhkhfgj?amount=0.2345&awaitingConfirm=true

Then if you scan that URI, the wallet would force priority fees to make the best attempt at confirming in the next block.

**Comment by voisine** [comment]
Hi @keo, we definitely need to move forward on a solution. 8-16 hours is certainly a degenerate case, we haven't seen anything like that except during one volume surge several day ago. An hour is more typical of what we're seeing right now. 

Merchants and other services that have requirements for confirmation need a way to communicate this to wallets.

Are you currently using BIP70, or just normal URIs ?

I suggest we create an extension for bip70 to allow services like yours to indicate the fees they are willing to cover and the minimum fee rate they expect.

**Comment by AndersBNielsen** [comment]
This is starting to be more than just a slight annoyance.
Blockchain.info considers breadwallet transactions high priority - but I've been waiting 9 hours for two 1BTC transactions to confirm.. Still waiting.

I agree there should be a way to set higher fees manually. 

I don't believe the correct move is to set higher fees for all transactions though - that'll just drive fees up unnecessarily.

Unless there's magically consensus for a jump to 4mb blocks overnight something needs to be done wallet wise.

**Comment by voisine** [comment]
@AndersBNielsen we've decided to bump the default fee up to since multiple hour confirmation times are longer than the typical experience we are targeting, and I just updated the default for the next app update.

With fixed blocksize cap, fees have to rise for everyone until people get priced out.

**Comment by AndersBNielsen** [comment]
Increasing the default fee doesn't seem like a good idea. It'll mean fees are too high 5-6 days of the week :/ 

If breadwallet is about leaving people in charge of their own funds, a hardcoded static fee pr byte seems... Opposite that.
An update every time the fee goes up per byte doesn't seem like a great idea either. 

Two options to keep people from going crazy:
1) Dynamic fees that work
2) Let people add X to the default fee when blocks are full all day.

**Comment by voisine** [comment]
We do now support BIP133 feefilter messages. It will take the median feefilter rate from connected peers and add 50% to that, however that won't take effect until most of the network upgrades to bitcoind 0.13

In the mean time, the fee bump is pretty conservative, 33784 sat-per-kb vs 26179.

We don't want users to have to understand how fees are calculated, or even what in the world a "confirmation" is in order to have a great experience using bitcoin. We really do need to push forward on letting merchants and other services indicate what kind of payments they will accept or not in order to render service. This is the only real solution. The recipient is the one who cares about confirmation, so they should be able to offer to pay for it.

**Comment by AndersBNielsen** [comment]
I disagree with the idea of obfuscating concepts that any new Bitcoin user will inevitably run into and any current user is already familiar with - people will notice their 10 cents missing... And they will absolutely notice their 1$ missing when the price rises and the fee rises at the same time.
But I guess that's a "vision"-issue.

That said, I'm very happy to hear Breadwallet will be usable again - I'm sure it will stay everyone's favorite iOS wallet :)

**Comment by trevorchill** [comment]
@voisine, futzing with the default fees or the algorithm is not enough. Please be gracious enough to understand when your app is not meeting the needs of your legitimate users in a practical way. 

Even if a user is a complete newbie to bitcoin, they are not being served when they discover that they can't send any transactions to get into the next block simply because you've hard coded it that way. Or that they have to wait hours and hours. Aaron, honestly, the app is failing us. We NEED the option to set a high enough fee to get to the top of the pile. 

After my experience the other day waiting hours for a confirmation when I could easily have added the fee, I'm considering going completely anti-breadwallet. You are verging on tanking your own app by being an ideologue forcing users to wait long times when it's completely unnecessary. What is the deal with refusing to add a setting for "use highest possible fees" or "use default fees"??  Your obstinacy has become unreasonable here.

**Comment by trevorchill** [comment]
Also, @voisine, your users are not getting "a great experience" with bitcoin now.  They already know they're paying a fee. Can't they be allowed to pay one that gives a good user experience?

And the idea that only merchants care about the delay is false. Please, let me emphasize this. Payments to ATMs, transfers to exchanges, payments in person that need a confirmation to be trusted, timeouts... -- the payor nearly always cares about delays, and very often is transferring to himself. The most critical time of all is when people want to cash out btc because of market movements and need to get to an exchange.

**Comment by dabura667** [comment]
I agree with the sentiment that "getting in the next block (or at least making an attempt to)" is the default behavior that people assume happens.

Yes, there are other factors at play here, and it would be nice if we could change those things.

However, the reality is that the network is going to have growing pains for the forseeable future, and users will switch wallets if they feel the problems are happening too often.

If congestion increases, and you notice less people using breadwallet, that would be a clear sign.

For the rest of us, if getting confirmed in the next block is important, use another wallet. Aaron has made it clear that breadwallet is targetting 1 hour (aka not even attempting to get into the next block) so, take it or leave it.

I am of the "always attempt at getting in the next block" camp personally.

**Comment by keo** [comment]
I agree with previous commenters that people using breadwallet and having pains already know about fees.

Sending it to an exchange, to an ATM, to a debit card, or someone else (met through Localbitcoins for example) and waiting for one or more confirmations is already a major use case.

Yes, this should be added to BIP70 but it should also be something that users can initiate too.

This is such a slight change in the UI (and can easily be hidden from novice users) that adding it would be a no brainer.

**Comment by billinghamj** [comment]
Perhaps the wallet should determine the current average spread of fees at the time of the payment, then offer the user the option to choose a low/medium/high fee, defaulting at medium? Low/medium/high could map to the 25th/50th/75th percentile of current fees.

That would be a good way of enabling it while maintaining good UX.

**Comment by paintballpanda** [comment]
Another data point -- we are a Bitcoin ATM operator and have been having the same issue. Users who are using breadwallet are facing difficulties when trying to sell their bitcoin to us.

Our customers who use the machine to cash out their bitcoins have been switching to different wallets so that they have the ability to control their fees. We love breadwallet and would like to continue recommending it but currently due to the fact that a user has no control over the fee setting we have been recommending alternatives.

I'm a little disappointed by the discussion here and how @voisine is offloading the responsibility to the payee. The payee is not the only party that has a real interest in determining the speed with which a transaction confirms. A payor has just as much interest in seeing their transaction confirm quickly. One such use case is our example of when a customer wants to receive their cash without having to wait for hours. 

@voisine I urge you to take a look at real world use cases and how breadwallet is failing those users rather than just sticking to your current course of not allowing a user to have any options whatsoever. If you continue to just offload all issues to another party so that you don't have to add any features to your wallet, users will quickly start to use alternatives.

**Comment by Transisto** [comment]
GOOD JOB !  Default fee of 26 sat. while recommended fee is above 100 sat/byte according to 21.co

https://tradeblock.com/bitcoin/tx/650a01ee166819804e575896fb1bb9d94cc79402bb663e61c60c2251cbddd72f

**Comment by voisine** [comment]
@Transisto that is not a breadwallet transaction, right? Currently we're at 65 sat/b

**Comment by Transisto** [comment]
Yes it was a breadwallet transaction. Maybe an older version.

**Comment by paintballpanda** [comment]
@voisine What is breadwallet currently at? It seems to be too low for transactions to confirm quickly enough.

https://blockchain.info/tx/795a254a54ab150d70d061f5b35488c4df3474309c80367eec237bda83481dba

125 sat/B is too low. At the same time 21.co recommends 220 sats/B for half-hour confirmation and 180 sats/B for 1 hour confirmation.

Can we have these fees bumped please or allow users the ability to send with a priority fee? As an ATM operator doing thousands of transactions a month, we are at the front lines and are the first to feel the impact of a wallet not keeping up with the ever increasing network fees.

**Comment by voisine** [comment]
@paintballpanda we're targeting 2-3 block typical confirmation times. In our own analysis 125s/B is providing that (as of the moment). We have to balance both timely confirmation with the fact that every fee increase causes smaller outputs to become un-spendable, and reduces the use cases that bitcoin is useful for, limiting it's adoption.

I appreciate your perspective, and we want breadwallet transactions to be useful for your use case as well which is why we're not going for the absolute lowest fees possible.

**Comment by paintballpanda** [comment]
@voisine do you have any plans to allow users the ability to select how small or large of a fee they would like to pay based upon how soon they need the transaction to confirm? It seems that by not giving users the choice inside breadwallet, not only are we limiting bitcoin's adoption but also breadwallet's adoption. 

The transaction I gave above as an example took approximately 2 hours 15 minutes to confirm. There really should be some option to allow users to target next block confirmation if they need to.

**Comment by overheadapproach** [comment]
@voisine Shouldn't users have the choice to decide how competitive to make their fee? Without this choice will a "fee market" ever develop?

**Comment by Transisto** [comment]
Just did a transaction between my own wallets, the cheapest low priority
fee was 146 sat/byte. Maybe that could be an indication of the normal fee
for bread.

**Comment by todb-r7** [comment]
@voisine can i suggest renaming this issue to:

**User-selectable fee**

Or somesuch? Right now, the issue name is a complaint, and not really a request, and it seems the fiddling with the fee schedule in a hardcoded way is a very slow solution.

There are a million ways to do it, but just acknowledging the request *as a request* might be a good start to get a contributor in the right headspace.

**Comment by paintballpanda** [comment]
@voisine do you have any plans to allow users the ability to select how small or large of a fee they would like to pay based upon how soon they need the transaction to confirm? It seems that by not giving users the choice inside breadwallet, not only are we limiting bitcoin's adoption but also breadwallet's adoption.

The transaction I gave above as an example took approximately 2 hours 15 minutes to confirm. There really should be some option to allow users to target next block confirmation if they need to.

I really hope breadwallet devs don't believe that they know better than the user, for all users and for all use cases.

**Comment by voisine** [comment]
@todb-r7, thanks for the suggestion

@paintballpanda The users we're targeting often have a difficult enough time grasping that USD and bitcoin are different currencies. 

Trying to get users to understand what a confirmation is and why they should want it is a level of complexity that's a bridge too far. Our goal is to make using bitcoin as simple and safe as possible. Confirmation times are random and unpredictable and a ~2 hour confirmation is something that can happen on occasion no matter how high a fee you pay. Users shouldn't have to care about it, and that's the goal we're working toward. 

The one-size-fits-most solution we have now is not perfect, I admit. We're definitely making a sacrifice to hide this complexity from the user, and not everyone will agree that it's the right trade-off, especially if their needs are different from the majority of our users. There are solutions however, from having expected fee data provided in payment requests, future 0-conf protocol layers like lightning, or multi-sig services like bitgo instant. That's the direction we're pursuing.

**Comment by paintballpanda** [comment]
@voisine Running a large bitcoin ATM network, we find ourselves catering to the same customers that you are targeting.

While many of these customers do indeed have difficulty understanding the technical details of bitcoin, they do want to be able to send and receive their bitcoin in a timely manner.

While confirmation times are indeed random, we do know that higher fees result, in most cases, in shorter wait times. Yes the occasional ~2 hour confirmation time can take place regardless of the fee but that is a much easier scenario to explain to customers than explaining to them that whenever the network is congested (almost a daily occurrence) their wallet no longer works unless they are happy to wait 6+ hours for a confirmation.  As of 10 minutes ago, breadwallet is still sending with a 175 sat/B fee whereas 21.co recommends a fee of 420 sat/B for confirmation within 1 hour. There is a very minute likelihood that a 175 sat/B fee will result in a 2 hour confirmation time. 

If you are sincerely interested in making bitcoin "as simple and safe as possible" then trying to kick the can down the road of technical understanding is not a sustainable methodology. Shying away from explaining to your customers what confirmation times and blocks are and how they affect their use of bitcoin is not making it simple nor safe for them to use bitcoin. The safety behind bitcoin is so intrinsically linked to blocks and confirmation times that trying to hide that from them in the hopes of delivering a safer or simpler user experience ends up delivering neither. 

I find it odd that you were so intent on forcing customers to use bits as units when that was not standard practice anywhere else -- on that topic you were very intent on having customers learn something because you held an opinion (which still seems to be the minority) that bits was the right unit to use and that although it may be difficult in the beginning, it would be beneficial for them and the ecosystem if they were to overcome the mental inertia and got used to it. 

Now with blocks and confirmation times you'd rather have customers not understand the very basic fundamentals of what is going on? We know the struggle--we are more on the front lines than you are and we know that explaining these concepts to a customer base that has little to no technical background is not the easiest thing to do. But we also know that the delayed cost of not explaining the underlying principles of how bitcoin works is not doing our customers or ourselves any favors.  

Understanding blocks and confirmation times seems a significantly more important lesson to learn than understanding an arbitrary unit of measure and I am surprised that you would be so insistent on users understanding one of these concepts and so resistant on them understanding the other.

While we cannot force you to change your mind, perhaps hearing the frustrations of your target market that we hear daily may help you realize the effect that your code is having on thousands of people. To this end, we have instructed our customer service staff to encourage breadwallet users to email your company's support directly with any questions or comments that they have regarding their delayed transactions.

**Comment by voisine** [comment]
@paintballpanda I don't understand what the bits denomination has to do with this. Bits are easier to understand than full btc (unless you've already gotten used to btc) Switching the industry to bits makes bitcoin easier to use.

Making the sender responsible for understanding how confirmations work makes bitcoin harder to use. They shouldn't have to understand it any more than they understand how credit card settlement happens. It doesn't follow that just because something is important, you must understand it in order to use it.

To that end, does your ATM network use the BIP70 payment protocol? We could extend the payment protocol to include a fee-per-kb field that your ATMs recommend for confirmation. This would solve the problem without requiring the user to understand anything. The fee just works, and you as the business operator get to determine what level is best for your customer experience.

**Comment by overheadapproach** [comment]
@voisine The issue isn't with people sending bitcoin to a terminal. It has to do with people wanting to spend the bitcoin in their wallet - bitcoin they bought from the BATM.

Right now when someone uses Breadwallet their transaction is broadcast and stuck in limbo for up to 2 weeks. It is a terrible user experience. They need to have the option to spend more on their fee if they want to have their transaction confirmed in a timely manner. What good is bitcoin to someone if they can't spend it?

We can't all use the same fee. Bitcoin will become a lottery system where only the lucky get confirmations. Users must be able to increase their fee if we are going to exist in a world with a finite amount of block space.

At this point users of bitcoin must be made aware that they are competing with all the other users for limited block space. There is no other way to explain it. It is the reason why.

**Comment by voisine** [comment]
If a transaction is stuck for 2 weeks, that's clearly unacceptable. We increase wallet fees to prevent that from happening, as we did earlier today. If you get reports of stuck transactions, please do direct them to support@breadwallet.com. That shouldn't happen, and if it does, we messed up.

The current system is not a lottery, the fee will continue to increase until it gets expensive enough that people choose not to use it (or their funds just become unspendable). It's not true that users need to know why (although its certainly useful to know). Just like when gas prices increase, people drive less even if they don't know about the geopolitical events that caused it.

**Comment by overheadapproach** [comment]
@voisine 2 weeks is now the default amount of time that a Core node retains transactions in its mempool. Used to be 72 hours. If breadwallet sends an uncompetitive fee, and the block space market does not become less competitive due to reduced demand in the short-term, then that is the wait time for a customer to regain control over their funds. The options are either get it there in a timely manner, or potentially offer the worst customer experience possible. How do you explain that to a customer without mentioning block space and miner fees? Where do you say that the customer's funds are and what do you say will happen to them?

If we all are using the same fee then it DOES become a lottery. How else will miners pick between our transactions but by random chance? What criteria might they use if they are not including the highest fees first? Perhaps it will be a first-come first-serve situation? I agree that some will choose to no longer use bitcoin, or no longer have spendable balances, but that will only increase your chances of winning the lottery if you remain.

Users demand to know why. It is their money and it neither exists here nor there in their eyes. People aren't complaining that sending BTC is expensive. They are complaining their money is gone.

**Comment by dabura667** [comment]
Fees are hard.

Today, my bitcoind bugged out and gave me 5849 satoshis per byte for nBlocks = 2 estimatefee call.

Almost sent a $500 fee... (scary)

**Comment by trevorchill** [comment]
The right way to handle it is by checking the prevailing fees and choosing a reasonable one dynamically whenever possible. 

This is what copay does. This is what most other wallets do now. There's really no other good option. 

Regarding difficulty for the user, there is none if you do it this way. Just have a setting like these other wallets in the advanced settings that allows you to choose a fee level. I am pretty much going to stop using breadwallet because of this, which is unfortunate.

**Comment by notDavid** [comment]
Hi, not sure if this is the right thread to comment on (if not feel free to delete this comment);

I just paid a €1,40 fee in Breadwallet (iOS) for a payment of €5,50. So the fee is about 25% of the paid amount. (just a small amount obviously.) Is this considered "normal" when paying in Bitcoins nowadays?

I didn't care about how fast the transaction would be confirmed, but no way to decrease the fee.

**Comment by dabura667** [comment]
https://bitcoinfees.21.co/ This site says the cheapest possible Bitcoin transaction should be €1,79

breadwallet is cheaper than most other fee estimates... and because of that breadwallet tends to take slightly longer than other wallets.

**Comment by tugruler** [comment]
Send 35$ and the fee was 9.77$ which is stupid 
I am new to btc  I naively assumed the fee would be reasonable but it wasn't the case I will not be using it again , the app is nice though :)

**Comment by dabura667** [comment]
@tugruler That is the bitcoin networks fault, not breadwallet's.

The more transactions you receive the higher your fees will be. Sending transactions will lower the fees for the next transaction by consolidating.

Fees have nothing to do with the sending amount.

**Comment by voisine** [comment]
@tugruler the fee is based on how large your transaction is in bytes. If you receive several smaller transactions and then spend them all at once, the fee will be several multiples larger than if you receive a single large amount and spend that.

**Comment by tugruler** [comment]
I was missing 30 cents worth of btc for a payment I went to ATM got some btc and made the payment  I assume it's because the deposit wasn't confirmed before I send the money and results as high fee ? 
As I said I am new to it and I don't speak bytes:)

**Comment by tugruler** [comment]
And I said nice app since I don't have any problem with the app itself :)

**Comment by xapple** [comment]
-> You need to let a way for the user to set the fee. <-
Some transactions need to go fast, others are non-urgent and would benefit from a smaller loss.
Users will not stick with bread in the long run if you can't customize that fee.

**Comment by xapple** [comment]
If anyone is looking for a way to send a transaction with a custom user-set fee, you can easily input the 12 word seed from bread wallet into this other wallet that gives you a bit more freedom: multibit.org

Just select the "Lighthouse (BIP 32)" option when inputing the 12 word seed.

**Comment by samuraisam** [comment]
Please don't go around inputting your seed into random websites. Follow the advice and never share your seed with anyone and especially never type it into a computer. We are working on improving the fee situation but the reality is that fees are still high. If you really need to move to another wallet do so by sending your balance rather than treating the 12 word phrase as a portable private key. It's not.

**Comment by paintballpanda** [comment]
Or maybe we should just let users be able to select their own fee?

**Comment by samuraisam** [comment]
As I said, we are working on improving the fee situation and fee selection is amongst the options that we are considering.

**Comment by lukerohde** [comment]
To set a priority would be great even if buried in settings . I transferred into bread wallet thinking it'd make everyday transactions quick and easy but the fees are too high for this. I just transferred out of breadwallet for small, low priority transactions. 

When dealing with bitcoin newbies like me its important to educate them about inescapable bitcoin idiosyncrasies rather than have them rage out in confusion for days as they work these fee issues out on their own (like I did). 

Having a priority setting would indicate that my transaction is competing with other transactions to clear.  I can pay to get to the head of the que or I can be stingy and suffer delays. 

In Australia, regular bank transfers  take a couple days to clear, or it costs a couple dollars to  instantly  draw cash out of a different bank's ATM.  Aussies are already used to these financial annoyances and choose a transaction mode that makes sense. Bitcoin has this flexibility so make it a strength not a weakness. 

Another option would be to set a max percentage. Spending $1.60 to clear a $5.00  transaction makes no sense especially when smaller transaction probably don't need to clear quickly. I don't like this idea as much since it forces small transactions to be slow and big ones to be expensive and fast. 

I understand your mission of making  bitcoin accessible to everyday folks by trying to hide it's complexity but when you hide the wrong stuff you drive your users away.  Good luck striking the balance!

**Comment by fubsier** [comment]
I have 4 transactions pending on the blockchain because of Bread Wallets stupidly low fees... They've been unverified for days and if transactions on the BTC blockchain remain this frequent there's no reason i can see why they won't stay unverified forever, and it would seem https://bitcoinfees.21.co/ webiste agrees with me... 
Bread Wallet needs to make their network fees higher ASAP. With BTC transactions increasing in volume more and more every day it makes no sense to keep fees this low and have your customers transactions put at the bottom of the pile... and further more, not to warn anyone about it. I'm so frustrated and will be telling everyone I know to avoid this wallet at all costs.

**Comment by dabura667** [comment]
breadwallet is trying to save users a tiny bit of money (the difference between 220 sat/byte and 190 sat/byte is 13% lower fees... but on average that would be going from $2 to $1.75...

I would 9 times out of 10 rather pay $2 over $1.75 to have a 70% higher chance of getting in the next block or two.

The common response of "block discovery times are random, so there are no guarantees" is almost useless in this case:

1. https://jochen-hoenicke.de/queue/#2h visualizing the mempool in this way is easier. Every time a block is found it clears transactions mostly from the top. and wallets are all aiming for the upper fee tiers to beat out the other people into the block.
2. If you aim for 4-5 tiers lower than what each block is clearing down to, you need to wait until many blocks come in quick succession, or else they will never dig down to your transaction in the fee pile.
3. Perhaps Aaron is holding this view point in a "if only every wallet was like breadwallet and stopped aiming for nBlocks=2 then fees wouldn't rise so quickly!" which harms users, as 99% of the "target user" (grandmas) sending solely to each other shouldn't care about 0 conf, 99% of "actual users" (aka people sending to and from services that need to wait for confirmations) would rather pay an extra 25 cents to get a +70% chance to get into the next block assuming you don't have bad luck and a dry spell of blocks pushed up the pile even higher.

The default should be "try your best to get into the next block, even if it's not guaranteed".

When I use Copay's "priority" (estimatefee nBlock=2) fee setting, I get into the next block almost every time. only 1 time I had to wait one block.
